### PR TITLE
feat: migrate to batch ingestion endpoint

### PR DESCRIPTION
**Pull request by maxdeichmann**

### CODE REVIEW

**Comment by maxdeichmann** [comment]
2 integration tests failing because this test takes the latest main docker server and is not able to cope with the new usage format. Succeeds locally.

File: langfuse-core/src/types.ts [hunk 1]
```diff
@@ -53,47 +56,99 @@ export type LangfuseMetadataProperties = {
...
-
-export const LangfusePostApiRoutes: Record<LangfuseObject, [LangfuseQueueItem["method"], keyof paths]> = {
-  createTrace: ["POST", "/api/public/traces"],
-  createEvent: ["POST", "/api/public/events"],
-  createSpan: ["POST", "/api/public/spans"],
-  updateSpan: ["PATCH", "/api/public/spans"],
-  createGeneration: ["POST", "/api/public/generations"],
-  updateGeneration: ["PATCH", "/api/public/generations"],
-  createScore: ["POST", "/api/public/scores"],
+export type CreateLangfuseTraceBody = {
```
Comment (marcklingen) [inline langfuse-core/src/types.ts:59]: could we load the types if we added them to the openapi spec?

File: langfuse-core/src/types.ts [hunk 2]
```diff
@@ -53,47 +56,99 @@ export type LangfuseMetadataProperties = {
...
+export type CreateLangfuseScoreBody = {
+  id?: string | null;
+  traceId: string;
+  name: string;
+  /** Format: double */
+  value: number;
+  observationId?: string | null;
+  comment?: string | null;
+};
+export type UpdateLangfuseSpanBody = {
```
Comment (marcklingen) [inline langfuse-core/src/types.ts:113]: I'd only specify what is different from the create here. can they be merged?

File: langfuse-core/src/types.ts [hunk 3]
```diff
@@ -53,47 +56,99 @@ export type LangfuseMetadataProperties = {
...
+  endTime?: Date | null;
+  name?: string | null;
+  metadata?: any | null;
+  input?: any | null;
+  output?: any | null;
+  level?: components["schemas"]["ObservationLevel"];
+  version?: string | null;
+  statusMessage?: string | null;
+};
+export type UpdateLangfuseGenerationBody = {
```
Comment (marcklingen) [inline langfuse-core/src/types.ts:128]: can events not be updated?

File: langfuse-core/src/index.ts [hunk 4]
```diff
@@ -93,7 +104,7 @@ abstract class LangfuseCoreStateless {
     this.publicKey = publicKey;
     this.secretKey = secretKey;
     this.baseUrl = removeTrailingSlash(options?.baseUrl || "https://cloud.langfuse.com");
-    this.flushAt = options?.flushAt ? Math.max(options?.flushAt, 1) : 1;
+    this.flushAt = options?.flushAt ? Math.max(options?.flushAt, 1) : 20;
```
[the remark this answers is not in the document]
Comment (marcklingen) [reply langfuse-core/src/index.ts:107]: means we set a lower bound to 1, we could go lower

File: langfuse-core/src/index.ts [hunk 5]
```diff
@@ -291,13 +303,17 @@ abstract class LangfuseCoreStateless {
...
+      this.flush((err, data) => {
+        return err ? reject(err) : resolve(data);
+      });
+    });
   }
 
   // Flushes all events that are not yet sent to the server
   // @returns {Promise[]} - list of promises for each item in the queue that is flushed
-  flush(): Promise<LangfuseFetchResponse>[] {
+  flush(callback?: (err?: any, data?: any) => void): void {
```
Comment (marcklingen) [inline langfuse-core/src/index.ts:316]: does it return promise or void?

File: langfuse-core/src/index.ts [hunk 6]
```diff
@@ -306,42 +322,41 @@ abstract class LangfuseCoreStateless {
...
+    const payload = JSON.stringify({ batch: items }); // implicit conversion also of dates to strings
+
+    const url = `${this.baseUrl}/api/public/ingestion`;
 
-    return promises;
+    const fetchOptions = this.getFetchOptions({
+      method: "POST",
+      body: payload,
+    });
+    console.debug("Sending payload", payload);
```
Comment (marcklingen) [inline langfuse-core/src/index.ts:351]: console

File: langfuse-core/src/converters.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  type CreateLangfuseGenerationBody,
+  type CreateLangfuseEventBody,
+  type Usage,
+} from "./types";
+
+export function convertEvent(body: CreateLangfuseEventBody): components["schemas"]["ObservationEvent"] {
+  const { id, traceId, startTime, level, ...rest } = body;
+
+  if (!id || !traceId) {
+    throw new Error("Cannot convert to ingestion format without id and traceId");
```
Comment (marcklingen) [inline langfuse-core/src/converters.ts:15]: throwing erros on optional attributes

File: langfuse-langchain/src/callback.ts [hunk 8]
```diff
@@ -123,6 +124,7 @@ export class CallbackHandler extends BaseCallbackHandler {
   async handleAgentAction(action: AgentAction, runId?: string, parentRunId?: string): Promise<void> {
     try {
       console.log("Agent action:", runId);
+
       this.langfuse.span({
```
Comment (marcklingen) [inline langfuse-langchain/src/callback.ts:128]: ._updateSpan() to only set end but not start?

**Pull request merged**